### PR TITLE
Implement Logging for Reviews

### PR DESCRIPTION
**Pull request by LouisJamesVasseur**
### What is this PR?

In this PR, logging is added to Reviews: each review is now connected to a User and a date, allowing for a better view of all the Reviews. Users will now only be able to submit 1 review per parking, which they can then edit by re-clicking on the Parking they added a review to

### Why? 

As Reviews become increasingly present in the app, we need a way to track the who/when of a review. Users now being able to edit their reviews also adds great functionality.

### How?

From the All Reviews Screen, users can now see if they've already added a Review for the Parking (visible on the FAB in the image below):

Screenshot 2024-11-10 at 17 22 51
Screenshot 2024-11-10 at 17 23 16

Text on both screens display "Edit" if User has already added a Review or "Add" if he hasn't.

### Testing

The added functionality has already been tested as a part of other PRs so no new tests are provided now

### Coverage Report

Screenshot 2024-11-10 at 17 21 14

### CODE REVIEW

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 1]
```diff
@@ -146,12 +146,21 @@ class ParkingViewModel(
...
+      isNewReview: Boolean
+  ) {
+    if (isNewReview) {
+      parking.avgScore =
+          (100 * ((parking.avgScore * parking.nbReviews) + newScore) / (parking.nbReviews + 1))
+              .toInt() / 100.00
+      parking.nbReviews += 1
+      parkingRepository.updateParking(parking, {}, {})
+    } else {
+      parking.avgScore += (oldScore - newScore) / parking.nbReviews
```
Comment (AlexandreBeeDeeBop) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:162]: I already did a similar comment on a previous PR of yours, I know it shouldnt happen but maybe create an edge case for parking.nbReviews = 0 just in case.

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 2]
```diff
@@ -31,23 +32,38 @@ import androidx.compose.ui.unit.dp
...
+import com.github.se.cyrcle.model.user.UserViewModel
 import com.github.se.cyrcle.ui.navigation.NavigationActions
+import com.github.se.cyrcle.ui.navigation.Screen
 import com.github.se.cyrcle.ui.theme.molecules.TopAppBar
+import com.google.firebase.Timestamp
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+fun Timestamp.toFormattedDate(): String {
+  val dateFormat = SimpleDateFormat("MMM dd, yyyy", Locale.getDefault())
```
Comment (AlexandreBeeDeeBop) [inline app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:44]: Typo. Also Timestamp.now() is doodoo i remember using it and it's pretty bad, small advice, dont rely on it, test it thoroughly !

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 3]
```diff
@@ -86,6 +100,21 @@ fun AllReviewsScreen(
                                           .testTag("ExpandedReviewBox$index")
                                           .clickable { setSelectedCardIndex(-1) }) {
                                     Column {
+                                      userViewModel.getUserById(curReview.owner)
+                                      Text(
+                                          text =
+                                              "Owner: ${userViewModel.currentUser.value?.username!!}",
```
Comment (AlexandreBeeDeeBop) [inline app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:106]: Can you really pretend "!!" ? Who knows if the user deletes his account... then it could become null.
Instead I think you could use "?: [DELETED_ACCOUNT]" or something like that.

Comment (AlexandreBeeDeeBop) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:106]: (Assuming the only reason a username is null is if the user was deleted)

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 4]
```diff
@@ -124,6 +152,22 @@ fun AllReviewsScreen(
                                             Modifier.fillMaxWidth()
                                                 .padding(16.dp)
                                                 .testTag("ReviewCardContent$index")) {
+                                          Text(
+                                              text =
+                                                  "Owner: ${userViewModel.currentUser.value?.username!!}",
```
Comment (AlexandreBeeDeeBop) [inline app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:157]: same here

File: app/src/main/java/com/github/se/cyrcle/ui/review/ReviewScreen.kt [hunk 5]
```diff
@@ -67,12 +83,12 @@ fun ReviewScreen(
               Text(
                   text =
                       when (sliderValue) {
-                        in 0f..1f -> stringResource(R.string.review_screen_terrible_review)
-                        in 1f..2f -> stringResource(R.string.review_screen_poor_review)
-                        in 2f..3f -> stringResource(R.string.review_screen_bad_review)
-                        3f -> stringResource(R.string.review_screen_average_review)
-                        in 3.5f..4f -> stringResource(R.string.review_screen_good_review)
-                        in 4.5f..5f -> stringResource(R.string.review_screen_great_review)
+                        in 0f..1f -> "Terrible experience!"
```
Comment (AlexandreBeeDeeBop) [inline app/src/main/java/com/github/se/cyrcle/ui/review/ReviewScreen.kt:86]: Why did you change from stringResource to string ?

**Review by AlexandreBeeDeeBop — COMMENTED** [review]
Overall, the "!!" are too risky for me to accept them. Using double bang operators is dangerous as it can cause NullPointerExceptions. It's better to provide default values / handle nulls gracefully.
Make sure you fix those style issues too, to please André who worked hard on them ;)

**Comment by AlexandreBeeDeeBop** [comment]
I might add that you writing "Users will now only be able to submit 1 review per parking" followed by a screenshot of "Owner: Louis V" having two reviews of a same parking spot is quite the ultimate ragebait.😂😂

**Review by AlexandreBeeDeeBop — APPROVED** [review]
Thanks for the changes

**Comment by Pijukai** [comment]
When you logIn as a guest you can add a review using your feature which go against what @matttsss did in his PR by updating the rule to prevent that.

**Comment by Pijukai** [comment]
When I click on see all reviews sometime the app just crashes i don't know if you have the same issue ?

File: app/src/main/java/com/github/se/cyrcle/model/review/Review.kt [hunk 6]
```diff
@@ -1,9 +1,12 @@
...
+import com.google.firebase.Timestamp
+
 data class Review(
     val uid: String,
     val owner: String,
     val text: String,
     val rating: Double,
-    val parking: String
+    val parking: String,
+    val time: Timestamp = Timestamp.now()
```
Comment (Pijukai) [inline app/src/main/java/com/github/se/cyrcle/model/review/Review.kt:11]: We agree that the user is not capable of choosing the time himself ?

**Review by Pijukai — APPROVED** [review]
Thanks for all the changes you have made. Seems like all the weird behaviors are gone good Job !
